src/Core/Managed/Shared/Extensibility/Metrics/MetricExtractorTelemetryProcessorBase.cs

```diff
@@ -0,0 +1,159 @@
+
+        protected MetricManager MetricManager
+        {
+            get { return _metricManager; }
+        }
+
+        protected virtual string ExtractorName
+        {
+            get
+            {
+                string name = _extractorName;
+                if (name == null)
+                {
+                    name = this.GetType().FullName;
+                    _extractorName = name;  // benign race
+                }
+                return name;
+            }
+        }
+
+        void ITelemetryModule.Initialize(TelemetryConfiguration configuration)
+        {
+            TelemetryClient telemetryClient = (configuration == null)
+                                                    ? new TelemetryClient()
+                                                    : new TelemetryClient(configuration);
+            if (MetricTelemetryMarkerKey != null)
+            {
+                telemetryClient.Context.Properties[MetricTelemetryMarkerKey] = MetricTelemetryMarkerValue;
+            }
+            
+            _metricManager = new MetricManager(new TelemetryClient(configuration));
+
+            Initialize(configuration);
+        }
+
+        void ITelemetryProcessor.Process(ITelemetry item)
+        {
+            if (item == null)
+            {
+                InvokeNextProcessor(item);
+                return;
+            }
+
+            try
+            {
+                Process(item);
+                AddExtractorInfo(item);
+            }
+            catch(Exception ex)
+            {
+                // Is this te right way to log erroes here?
+                CoreEventSource.Log.LogError("Error in Metric Extractor: " + ex.ToString());
+            }
+
+            InvokeNextProcessor(item);
+        }
+
+        private void AddExtractorInfo(ITelemetry item)
+        {
+            string extractorName = ExtractorName ?? "null";
+            string extractorVersion = ExtractorVersion ?? "null";
+
+            string thisExtractorInfo = $"(Name:{extractorName}, Ver:{extractorVersion})";
+
+            string extractionPipelineInfo;
+            bool hasPrevInfo = item.Context.Properties.TryGetValue(MetricTerms.Extraction.PipelineInfo.PropertyKey, out extractionPipelineInfo);
+
+            if (! hasPrevInfo)
+            {
+                extractionPipelineInfo = String.Empty;
+            }
+            else
+            {
+                if (extractionPipelineInfo.Length > 0)
+                {
+                    extractionPipelineInfo = extractionPipelineInfo + "; ";
+                }
+            }
+
+            extractionPipelineInfo = extractionPipelineInfo + thisExtractorInfo;
+            item.Context.Properties[MetricTerms.Extraction.PipelineInfo.PropertyKey] = extractionPipelineInfo;
+        }
+
+        private void InvokeNextProcessor(ITelemetry item)
+        {
+            ITelemetryProcessor next = _nextProcessorInPipeline;
+            if (next != null)
```

Comment:
I prefer the new form in assignments like x = foo?.ComputeValue(), but not in statements like do this when x isn't null. I think, both is generally fine, though. :)